template/upstream.c.erb

```diff
@@ -27,22 +27,63 @@ static void btx_dispatch_<%= e.name_sanitized %>(
   const bt_clock_snapshot *clock_snapshot = bt_message_event_borrow_default_clock_snapshot_const(upstream_message);
   bt_clock_snapshot_get_ns_from_origin(clock_snapshot, &_timestamp);
   <% end %>
-  // Call all the callbacks who where registered
-  // Their type are declared in 'upstream.h'
-  <%= e.name_sanitized %>_callback_f **p = NULL;
-  while ((p = utarray_next(callbacks->event_callbacks, p))) {
-    (*p)((void *)common_data,
-         common_data
-             ->usr_data<%= ', _timestamp' if e.default_clock_class %><%= e.args.map{ |s| s.name }.join_with_prefix(", ") %>);
+
+  /* Call registered event callbacks */
+  {
+    // The callbacks array is allocated if the user register a callback for it.
+    // If no callback was registered, we can not iterate over the array, since 
+    // it is a null pointer. 
+    if (callbacks->event_callbacks) {
+      void **p = NULL;
+      while ((p = utarray_next(callbacks->event_callbacks, p))) {
+        (*((<%= e.name_sanitized %>_callback_f **)(p)))((void *)common_data,
+            common_data
+                ->usr_data<%= ', _timestamp' if e.default_clock_class %><%= e.args.map{ |s| s.name }.join_with_prefix(", ") %>);
+      }
+    }
+  }
+
+  /* Call registered matching callbacks */
+  <% matching_dispatchers.each do |m| %>
+  <% if m.match?(e.name) %>
+  {
+      if (callbacks-><%= m.name_sanitized %>_callbacks) {
+      void **p = NULL;
+      while ((p = utarray_next(callbacks-><%= m.name_sanitized %>_callbacks, p))) {
+        (*((<%= m.name_sanitized %>_callback_f **)(p)))((void *)common_data,
+            common_data
+                ->usr_data<%= m.args.map{ |s| s.name }.join_with_prefix(", ") %>);
+      }
+    }
```

Comment:
This can be merged in the same loop.
The "event_callancks" is `matching` callbacks like any other.